packages/vite-plugin-cloudflare/src/index.ts

```diff
@@ -407,67 +410,53 @@ export function cloudflare(pluginConfig: PluginConfig = {}): vite.Plugin[] {
 				// Skip this whole plugin if we are in preview mode
 				return !env.isPreview;
 			},
-			config() {
-				// Configure Vite with the Node.js polyfill aliases
-				// We have to do this across the whole Vite config because it is not possible to do it per Environment.
-				// These aliases are to virtual modules that then get Environment specific handling in the resolveId hook.
+			applyToEnvironment(environment) {
+				const workerConfig = getWorkerConfig(environment.name);
+				return isNodeCompat(workerConfig);
+			},
+			// We need the resolver from this plugin to run before built-in ones, otherwise Vite's built-in
+			// resolver will try to externalize the Node.js module imports (e.g. `perf_hooks` and `node:tty`)
+			// rather than allowing the resolve hook here to alias then to polyfills.
+			enforce: "pre",
+			configEnvironment() {
 				return {
 					resolve: {
-						alias: getNodeCompatAliases(),
+						builtins: getNodeCompatExternals(),
+					},
+					optimizeDeps: {
+						// We don't want to follow any `node:*` type imports when optimizing deps,
+						// whether or not they are built-in to workerd or polyfilled here.
+						// The ones that are not built-in will just get resolved to their polyfill on demand.
+						exclude: [...nodeBuiltInModules],
+						// But we do want to ensure that all the possible entry-points for the polyfills get pre-bundled
+						// so that they are ready for the first request to the dev server.
+						include: [...getNodeCompatEntries()],
```

Comment:
Just checking that it's been confirmed across package managers that ESBuild can resolve the unenv imports here? Do the new E2E tests for the plugin run on every PR?

Reply:
The new e2e tests run on every PR - as part of the `test:ci` run for the `@cloudflare/vite-plugin` package.
I will also double check with a manually created project.

Reply:
Tested manually and all seems good.